### PR TITLE
bpf: xdp: clean up xdp_adjust_hroom()

### PR DESCRIPTION
**Pull request by julianwiedmann**
Clean up some obscure details in `xdp_adjust_hroom()`, so that the next traveller on this road will have an easier journey.

### CODE REVIEW

File: bpf/include/bpf/ctx/xdp.h [hunk 1]
```diff
@@ -319,16 +315,23 @@ ctx_adjust_hroom(struct xdp_md *ctx, const __s32 len_diff, const __u32 mode,
...
 			break;
 		default:
 			__throw_build_bug();
 		}
+
+		if (move_len) {
+			void *data_end = ctx_data_end(ctx);
+			void *data = ctx_data(ctx);
+
+			if (data + move_len + len_diff <= data_end)
```
Comment (gentoo-root) [inline bpf/include/bpf/ctx/xdp.h]: Nit: `data + len_diff + move_len` would be a more natural order, IMO (offset + length).

File: bpf/include/bpf/ctx/xdp.h [hunk 2]
```diff
@@ -291,25 +290,22 @@ ctx_adjust_hroom(struct xdp_md *ctx, const __s32 len_diff, const __u32 mode,
 	 * this must be made more generic.
 	 */
 	if (!ret) {
-		data_end = ctx_data_end(ctx);
-		data = ctx_data(ctx);
+		__u32 move_len = 0;
+
+		/* Based on the specified `len_diff`, we now *guess* at what
+		 * offset inside the packet the free space should be created.
+		 * If offset > 0, we need to move some packet data forward.
```
Comment (gentoo-root) [inline bpf/include/bpf/ctx/xdp.h]: > If offset > 0

What is offset? Does it refer to some variable in the code? I don't see one with that name.

> we need to move some packet data forward.

But what __bpf_memmove_fwd actually does is move the headers backward (to the newly allocated headroom), not move the payload forward.

I understand what's going on in the function, but that sentence in the comment sounds confusing. Could you rephrase it somehow? Maybe it's worth elaborating that there are two cases: in one case we prepare for insertion of new headers before the existing ones (`move_len == 0`), and in the other we want to squeeze in a new header after the existing ones (`move_len > 0`), so we are moving the old headers in the latter case.

Comment (julianwiedmann) [reply bpf/include/bpf/ctx/xdp.h]: Ty - that's good feedback, I pushed an adjusted version. I think we might have different ideas where the front/back of a packet is :grin:.

Comment (gentoo-root) [reply bpf/include/bpf/ctx/xdp.h]: Thanks! I like the new version.

**Review by gentoo-root — APPROVED** [review]
